File: problem/Maximize_Subarrays/maxSubarrays.py
class Solution:
    def maxSubarrays(self, n, conflictingPairs):
        conflicts = [[] for _ in range(n + 2)]

        for a, b in conflictingPairs:
            if a < b:
                conflicts[b].append(a)
            else:
                conflicts[a].append(b)
        
        max_left = 0
        second_max_left = 0
        gains = [0] * (n + 2)
        valid_subarrays = 0

        for right in range(1, n + 1):
            for left in conflicts[right]:
                if left > max_left:
                    second_max_left = max_left
                    max_left = left
                elif left > second_max_left:
                    second_max_left = left
            valid_subarrays += right - max_left
            gains[max_left] += max_left - second_max_left

        return valid_subarrays + max(gains)

File: problem/Maximize_Subarrays/maxSubarrays.java
class Solution 
{
    public long maxSubarrays(int n, int[][] conflictingPairs) 
    {
        List<Integer>[] conflicts = new List[n + 2];
        for (int i = 0; i < conflicts.length; i++) 
        {
            conflicts[i] = new ArrayList<>();
        }
        
        for (int[] pair : conflictingPairs) 
        {
            int a = pair[0];
            int b = pair[1];
            
            if (a < b) 
            {
                conflicts[b].add(a);
            } 
            else 
            {
                conflicts[a].add(b);
            }
        }

        int maxLeft = 0;
        int secondMaxLeft = 0;
        long[] gains = new long[n + 2]; 
        long validSubarrays = 0;

        for (int right = 1; right <= n; right++) 
        {
            for (int left : conflicts[right]) 
            {
                if (left > maxLeft) 
                {
                    secondMaxLeft = maxLeft;
                    maxLeft = left;
                } 
                else if (left > secondMaxLeft) 
                {
                    secondMaxLeft = left;
                }
            }

            validSubarrays += right - maxLeft;
            gains[maxLeft] += maxLeft - secondMaxLeft;
        }

        long maxGain = 0;
        for (long gain : gains) 
        {
            if (gain > maxGain) 
            {
                maxGain = gain;
            }
        }

        return validSubarrays + maxGain;
    }
}